pkg/agent/multicluster/mc_route_controller.go

```diff
@@ -214,40 +250,233 @@ func (c *MCRouteController) Run(stopCh <-chan struct{}) {
+// function in order to read and process a message on the gatewayQueue.
+func (c *MCRouteController) gatewayWorker() {
+	for c.processNextGatewayItem() {
 	}
 }
 
-func (c *MCRouteController) processNextWorkItem() bool {
-	obj, quit := c.queue.Get()
+// ciImpWorker is a long-running function that will continually call the processNextClusterInfoImportItem
+// function in order to read and process a message on the ciImpQueue.
+func (c *MCRouteController) ciImpWorker() {
+	for c.processNextClusterInfoImportItem() {
+	}
+}
+
+func (c *MCRouteController) processNextClusterInfoImportItem() bool {
+	obj, quit := c.ciImpQueue.Get()
 	if quit {
 		return false
 	}
-	defer c.queue.Done(obj)
+	defer c.ciImpQueue.Done(obj)
 
-	if k, ok := obj.(string); !ok {
-		c.queue.Forget(obj)
+	ciImp, ok := obj.(string)
+	if !ok {
+		c.ciImpQueue.Forget(obj)
 		klog.InfoS("Expected string in work queue but got", "object", obj)
 		return true
-	} else if err := c.syncMCFlows(); err == nil {
-		c.queue.Forget(k)
+	}
+	if c.wireGuardConfig != nil {
+		if err := c.updateWireGuardPeers(ciImp); err != nil {
+			c.ciImpQueue.AddRateLimited(ciImp)
+			klog.ErrorS(err, "Error updating WireGuard peers, requeuing", "key", ciImp)
+		}
+	}
+
+	if err := c.syncMCFlows(); err == nil {
+		c.ciImpQueue.Forget(obj)
+	} else {
+		// Put the item back on the workqueue to handle any transient errors.
+		c.ciImpQueue.AddRateLimited(ciImp)
+		klog.ErrorS(err, "Error syncing key, requeuing", "key", ciImp)
+	}
+	return true
+}
+
+func (c *MCRouteController) processNextGatewayItem() bool {
+	obj, quit := c.gatewayQueue.Get()
+	if quit {
+		return false
+	}
+	defer c.gatewayQueue.Done(obj)
+
+	gateway, ok := obj.(string)
+	if !ok {
+		c.gatewayQueue.Forget(obj)
+		klog.InfoS("Expected string in work queue but got", "object", obj)
+		return true
+	}
+
+	if c.wireGuardConfig != nil {
+		if err := c.reconcileWireGuard(); err != nil {
+			klog.ErrorS(err, "Error when reconcile WireGuard, requeuing", "key", gateway)
+			c.gatewayQueue.AddRateLimited(gateway)
+		}
+	}
+	if err := c.syncMCFlows(); err == nil {
+		c.gatewayQueue.Forget(obj)
 	} else {
 		// Put the item back on the workqueue to handle any transient errors.
-		c.queue.AddRateLimited(k)
-		klog.ErrorS(err, "Error syncing key, requeuing", "key", k)
+		c.gatewayQueue.AddRateLimited(gateway)
+		klog.ErrorS(err, "Error syncing key, requeuing", "key", gateway)
 	}
 	return true
 }
 
+func (c *MCRouteController) reconcileWireGuard() error {
+	gateways, err := c.mcClient.MulticlusterV1alpha1().Gateways(c.namespace).List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+	if c.wireGuardInitialized {
+		// The Gateway is deleted. Remove WireGuard interface
+		if len(gateways.Items) == 0 || gateways.Items[0].Name != c.nodeConfig.Name {
+			klog.InfoS("Gateway is deleted, removing WireGuard interface and rule")
+			if err := c.deleteWireGuardInterfaceAndRule(); err != nil {
```

Comment:
Could you check if deleteWireGuardInterfaceAndRule supports reentrance? For example, DeleteIPRule should not return an error, when it is called the 2nd time at error retry.

Better to add a comment to `deleteWireGuardInterfaceAndRule` to indicate it needs to support reentrance.

Reply:
Add a check before delete now. So it won't return an error then delete it 2nd time.

Reply:
Could you add comments to all funcs that need to support reentrance, including `deleteWireGuardInterfaceAndRule` and `DeleteIPRule`?

> Better to add a comment to deleteWireGuardInterfaceAndRule to indicate it needs to support reentrance.

> Please revisit all these cases and probably add a comment to the funcs need to support reentrance.